olp-cpp-sdk-core/tests/thread/ThreadPoolTaskSchedulerTest.cpp

```diff
@@ -126,3 +129,129 @@ TEST(ThreadPoolTaskSchedulerTest, MultiUserPush) {
   }
   push_threads.clear();
 }
+
+TEST(ThreadPoolTaskSchedulerTest, Prioritization) {
+  auto thread_pool = std::make_shared<ThreadPool>(1);
+  TaskScheduler& scheduler = *thread_pool;
+
+  struct MockOp {
+    MOCK_METHOD(void, Op, (olp::thread::Priority));
+  } mockop;
+
+  testing::Sequence sequence;
+
+  EXPECT_CALL(mockop, Op(olp::thread::HIGH)).Times(10).InSequence(sequence);
+  EXPECT_CALL(mockop, Op(olp::thread::NORMAL)).Times(10).InSequence(sequence);
+  EXPECT_CALL(mockop, Op(olp::thread::LOW)).Times(10).InSequence(sequence);
+
+  std::mutex mutex;
+  std::condition_variable cv;
+
+  scheduler.ScheduleTask(
+      [&]() {
+        std::unique_lock<std::mutex> lock(mutex);
+        cv.wait(lock);
+      },
+      std::numeric_limits<uint32_t>::max());
+
+  const uint32_t expected_tasks = 30u;
+  uint32_t counter(0u);
+
+  const olp::thread::Priority priorities[] = {
+      olp::thread::LOW, olp::thread::NORMAL, olp::thread::HIGH};
+
+  for (uint32_t i = 0; i < expected_tasks; ++i) {
+    auto priority = priorities[i % 3];
+    scheduler.ScheduleTask(
+        [&, priority]() {
+          counter++;
+          mockop.Op(priority);
+        },
+        priority);
+  }
+
+  cv.notify_all();
+
+  const auto start = chrono::system_clock::now();
+  auto check_condition = [&]() {
+    return counter < expected_tasks &&
+           chrono::duration_cast<chrono::seconds>(chrono::system_clock::now() -
+                                                  start) < chrono::seconds(10);
+  };
+
+  while (check_condition()) {
+    std::this_thread::sleep_for(kSleep);
+  }
```

Comment:
This is not a very good system as it created a lot of flakiness. Instead, why not push a last, very low priority task that sets a promise<void> and you just wait for the associated future?